proposal/0000-record-dot-syntax.md

```diff
@@ -12,77 +12,276 @@ the link, and delete this bold sentence.**
 
 ## Proposed Change Specification
 
-Specify the change in precise, comprehensive yet concise language. Avoid words
-like "should" or "could". Strive for a complete definition. Your specification
-may include,
+For this extension we focus on the changes to the parsing rules, and the desugaring, with the expectation the type checking and renamer follow directly from that. To confirm these changes integrate as expected we have written [a prototype implementation](https://gitlab.haskell.org/shayne-fletcher-da/ghc/commits/record-dot-syntax) that parses and desugars the forms directly in the parser.
+
+### `RecordDotSyntax` language extension
+
+This change adds a new language extension (enabled at source via `{-# LANGUAGE RecordDotSyntax #-}` or on the command line via the flag `-XRecordDotSyntax`).
+
+When `RecordDotSyntax` is in effect, the use of '.' to denote function composition is disambiguated from the use of '.' to denote record field access by the presence (, respectively absence,) of whitespace trailing the '.'.
+
+Suppose the following datatype declarations.
+
+```haskell
+data Foo = Foo {foo :: Bar}
+data Bar = Bar {bar :: Baz}
+data Baz = Baz {baz :: Quux}
+data Quux = Quux {quux :: Int}
+```
+
+The existence of the builtin `HasField` typeclass means that it is possible to write code for getting and setting record fields like this:
+
+```haskell
+getQuux :: Foo -> Int
+getQuux a = getField @"quux" (getField @"baz" (getField @"bar" (getField @"foo" a)))
+
+setQuux :: Foo -> Int -> Foo
+setQuux a i = setField@"foo" a (setField@"bar" (getField @"foo" a) (setField@"baz" (getField @"bar" (getField @"foo" a)) (setField@"quux" (getField @"baz" (getField @"bar" (getField @"foo" a))) i)))
+```
+
+`RecordDotSyntax` enables new concrete syntax so that the following program is equivalent.
+
+```haskell
+getQuux a = a.foo.bar.baz.quux
+setQuux a i = a{foo.bar.baz.quux = i}
+```
+
+In the event the language extension is enabled:
+
+| Expression | Equivalent |
+| -- | -- |
+| `e.lbl` | `getField @"lbl" e` the `.` cannot have whitespace on either side |
+| `e{lbl = val}` | `setField @"lbl" e val` |
+| `(.lbl)` | `(\x -> x.lbl)` the `.` cannot have whitespace after |
+| `e{lbl1.lbl2 = val}` | `e{lbl1 = (e.lbl1){lbl2 = val}}` performing a nested update |
+| `e{lbl * val}` | `e{lbl = e.lbl * val}` where `*` can be any operator |
```

Comment:
IMO, this _concrete_ syntax does not properly reflect that there's an "assignment" going on here. Can we take inspiration from C's (or Rust's) `*=` family of "operation + assignment" operators?

Reply:
I do have sympathy for what you say but dispute that the syntax does not reflect that there is a (functional) update `r{...}` - the leading `r` immediately designates an update expression. Of course in C (Rust), `i += 2` *modifies* `i`. That is of course "true" assignment for which there is no equivalent here (of course) and I think therefore adopting `op=` notation would be an abuse.

Reply:
This conversation is actually convincing me harder on the current syntax. Again, referencing the `i += 2` example, the target of the assignment is `i`. In `r{foo.bar + 2`}, the target of the update is `r`. In fact, it seems to me (ironically?) if we stick `with` in there and read the latter as `r with {foo. bar * 2}` it makes more and more sense to me :)

Reply:
Rohan tells me he's not convinced by my argument (following) but I offer it anyway as the current best articulation I've managed.

Why `r{foo.bar + 1}` and not `r{foo.bar += 1}`?

At first blush `r{foo.bar += 1}` looks appealing. It's not as good a fit as it first seems though. In C like languages, `i += 1` *assigns* to `i`. It is a statement - not an expression. It *affects* `i`. In pseudo Haskell, it's a shorthand for `i <- i + 1`.

As a C programmer, I want to read `r{foo.bar += 2}` as modifying `r`.

In Haskell, `r{...}` computes a new record from `r`. When we write `r{foo.bar + 1}` rather than `r{foo.bar += 1}` it is obvious neither `r` or `r.foo.bar` are affected.

It might help to read `r{foo.bar + 1}` as `r` *with* `foo.bar = r.foo.bar + 1`.

Reply:
Bernhard writes:

![Screen Shot 2019-10-03 at 3 49 44 PM](https://user-images.githubusercontent.com/41306216/66159204-7d183f00-e5f5-11e9-879b-f657017951af.png)

Reply:
Proposal supports:
  - `r{foo.bar ++ [1, 2, 3]}`
 - `r{foo.bar 'plus' 2}`
What would these look like under the `+=` scenario?

Reply:
Yep, I'm pretty convinced the proposed syntax is the cleanest option overall (of the two options anyway).

Reply:
I note that += is a totally valid Haskell operator, so would that look like `{a +== 1}`? What does `div`= look like?